src/lib/y2partitioner/dialogs/partition_size.rb

```diff
@@ -10,143 +13,335 @@ module Dialogs
-                  Id(:end_block),
-                  _("End Block"),
-                  min_block, max_block, end_block
-                )
+        # @macro seeAbstractWidget
+        def store
+          # nothing to do, that's OK
+        end
+      end
+
+      # Enter a human readable size
+      class CustomSizeInput < CWM::InputField
+        # @return [Y2Storage::DiskSize]
+        attr_reader :min_size, :max_size
+
+        def initialize(ptemplate, regions)
+          textdomain "storage"
+          @ptemplate = ptemplate
+          @regions = regions
+          largest_region = @regions.max_by(&:size)
+          @max_size = largest_region.size
+          @min_size = Y2Storage::DiskSize.new(1)
+        end
+
+        # Forward to ptemplate
+        def size
+          @ptemplate.custom_size
+        end
+
+        # Forward to ptemplate
+        def size=(v)
+          @ptemplate.custom_size = v
+        end
+
+        # @return [Y2Storage::Region] the smallest region
+        #   that can contain the chosen size
+        def parent_region
+          suitable_rs = @regions.find_all { |r| r.size >= size }
+          suitable_rs.min_by(&:size)
+        end
+
+        # @return [Y2Storage::Region] create it in the smallest region
+        #   that can contain the chosen size
+        def region
+          parent = parent_region
+          bsize = parent.block_size
+          length = (size.to_i / bsize.to_i.to_f).ceil
+          Y2Storage::Region.create(parent.start, length, bsize)
+        end
+
+        # @macro seeAbstractWidget
+        def label
+          _("Size")
+        end
+
+        # @macro seeAbstractWidget
+        def init
+          self.size ||= max_size
+          self.value = size
+        end
+
+        # @macro seeAbstractWidget
+        def store
+          self.size = value
+        end
+
+        # @macro seeAbstractWidget
+        def validate
+          v = value
+          if v.nil? || v > max_size
+            min_s = min_size.human_ceil
+            max_s = max_size.human_floor
+            Yast::Popup.Error(
+              Yast::Builtins.sformat(
+                # error popup, %1 and %2 are replaced by sizes
+                _("The size entered is invalid. Enter a size between %1 and %2."),
+                min_s, max_s
               )
             )
-          )
+            # TODO: Let CWM set the focus
+            Yast::UI.SetFocus(Id(widget_id))
+            false
+          else
+            true
+          end
         end
 
+        # @return [Y2Storage::DiskSize,nil]
+        def value
+          Y2Storage::DiskSize.from_human_string(super)
+        rescue ArgumentError
+          nil
+        end
+
+        # @param v [Y2Storage::DiskSize]
+        def value=(v)
+          super(v.human_floor)
+        end
+      end
+
+      # Specify start+end of the region
+      class CustomRegion < CWM::CustomWidget
+        def initialize(ptemplate, regions)
+          raise ArgumentError if regions.empty?
```

Comment:
can you add there message which describe what exactly is wrong with arguments? I think it will help with later debugging.